File: frontend/src/presenter.js
/**
 * Takes a raw resource object and returns a structured, human-readable summary.
 */
export function presentResource(resource) {
  const resourceType = resource?.resource_type?.toLowerCase();
  const data = resource?.data || {};

  const presenters = {
    'deployment': presentDeployment,
    'service': presentService,
    'secret': presentSecret,
    'ingress': presentIngress,
    'configmap': presentConfigMap,
    'pod': presentPod,
    'persistentvolume': presentPersistentVolume,
    'persistentvolumeclaim': presentPersistentVolumeClaim,
    'statefulset': presentStatefulSet,
    'daemonset': presentDaemonSet,
    'job': presentJob,
    'cronjob': presentCronJob,
    'horizontalpodautoscaler': presentHPA,
    'customresourcedefinition': presentCRD,
    'networkpolicy': presentNetworkPolicy,
  };

  const presenterFunc = presenters[resourceType];
  if (presenterFunc) {
    return presenterFunc(data);
  }

  return { "Info": "No specific summary view available for this resource type.", "Data": data };
}

function presentDeployment(data) {
  const spec = data.spec || {};
  const status = data.status || {};
  return {
    "Replicas": `${status.readyReplicas || 0} / ${spec.replicas || 0}`,
    "Strategy": spec.strategy?.type || 'N/A',
    "Containers": (spec.template?.spec?.containers || []).map(c => ({ Name: c.name, Image: c.image })),
    "Conditions": (status.conditions || []).map(c => `${c.type}: ${c.status}`),
  };
}

function presentService(data) {
  const spec = data.spec || {};
  return {
    "Type": spec.type,
    "Cluster IP": spec.clusterIP,
    "Ports": (spec.ports || []).map(p => `${p.name || ''} ${p.port}:${p.targetPort}/${p.protocol}`.trim()),
    "Selector": spec.selector || 'None',
  };
}

function presentSecret(data) {
  return { "Type": data.type, "Data Keys": Object.keys(data.data || {}) };
}

function presentIngress(data) {
    const spec = data.spec || {};
    return {
        "Ingress Class": spec.ingressClassName || 'None',
        "Rules": (spec.rules || []).map(rule => ({
            Host: rule.host,
            Paths: (rule.http?.paths || []).map(p => `${p.path} ➔ ${p.backend.service?.name}:${p.backend.service?.port?.number || p.backend.service?.port?.name}`),
        })),
    };
}

function presentConfigMap(data) {
    return data.data || {};
}

function presentPod(data) {
    const spec = data.spec || {};
    const status = data.status || {};
    return {
        "Status": status.phase,
        "Node": spec.nodeName,
        "IP": status.podIP,
        "Containers": (spec.containers || []).map(c => ({
            Name: c.name,
            Image: c.image,
            Ready: (status.containerStatuses?.find(cs => cs.name === c.name)?.ready ?? false).toString()
        }))
    };
}

function presentPersistentVolume(data) {
    const spec = data.spec || {};
    return {
        "Capacity": spec.capacity?.storage,
        "Access Modes": spec.accessModes,
        "Reclaim Policy": spec.persistentVolumeReclaimPolicy,
        "Status": data.status?.phase,
        "Claim": spec.claimRef?.name,
        "Storage Class": spec.storageClassName,
    };
}

function presentPersistentVolumeClaim(data) {
    const spec = data.spec || {};
    return {
        "Status": data.status?.phase,
        "Volume": spec.volumeName,
        "Capacity": data.status?.capacity?.storage,
        "Access Modes": spec.accessModes,
        "Storage Class": spec.storageClassName,
    };
}

function presentStatefulSet(data) {
    const spec = data.spec || {};
    return {
        "Replicas": `${data.status?.readyReplicas || 0} / ${spec.replicas || 0}`,
        "Service Name": spec.serviceName,
        "Update Strategy": spec.updateStrategy?.type,
    };
}

function presentDaemonSet(data) {
    const status = data.status || {};
    return {
        "Desired": status.desiredNumberScheduled,
        "Current": status.currentNumberScheduled,
        "Ready": status.numberReady,
        "Available": status.numberAvailable,
    };
}

function presentJob(data) {
    return {
        "Completions": data.spec?.completions,
        "Parallelism": data.spec?.parallelism,
        "Succeeded": data.status?.succeeded,
        "Failed": data.status?.failed,
    };
}

function presentCronJob(data) {
    return {
        "Schedule": data.spec?.schedule,
        "Suspend": data.spec?.suspend,
        "Last Schedule": data.status?.lastScheduleTime,
    };
}

function presentHPA(data) {
    const spec = data.spec || {};
    return {
        "Scale Target": `${spec.scaleTargetRef.kind}/${spec.scaleTargetRef.name}`,
        "Min Replicas": spec.minReplicas,
        "Max Replicas": spec.maxReplicas,
        "Current Replicas": data.status?.currentReplicas,
    };
}

function presentCRD(data) {
    const spec = data.spec || {};
    return {
        "Group": spec.group,
        "Scope": spec.scope,
        "Versions": (spec.versions || []).map(v => v.name),
    };
}

function presentNetworkPolicy(data) {
    return {
        "Pod Selector": data.spec?.podSelector,
        "Policy Types": data.spec?.policyTypes,
    };
}

File: utils/presenter.py
def get_structured_data(resource: dict) -> dict:
    """
    Parses a resource dictionary and returns a structured summary.
    """
    resource_type = resource.get("resource_type", "").lower()
    data = resource.get("data", {})

    if not data:
        return {"Error": "No data available for this resource."}

    presenter_functions = {
        "deployment": present_deployment,
        "service": present_service,
        "secret": present_secret,
        "ingress": present_ingress,
        "configmap": present_configmap,
    }

    presenter_func = presenter_functions.get(resource_type)
    if presenter_func:
        return presenter_func(data)

    return {"Info": "Standard presentation for this resource type."}

def present_deployment(data: dict) -> dict:
    """Presents a summary of a Deployment."""
    spec = data.get("spec", {})
    status = data.get("status", {})
    template_spec = spec.get("template", {}).get("spec", {})

    containers = template_spec.get("containers", [])
    container_info = [
        {
            "Image": c.get("image"),
            "Ports": [p.get("containerPort") for p in c.get("ports", [])],
        }
        for c in containers
    ]

    return {
        "Replicas": spec.get("replicas"),
        "Available Replicas": status.get("availableReplicas"),
        "Ready Replicas": status.get("readyReplicas"),
        "Strategy": spec.get("strategy", {}).get("type"),
        "Containers": container_info,
    }

def present_service(data: dict) -> dict:
    """Presents a summary of a Service."""
    spec = data.get("spec", {})
    return {
        "Type": spec.get("type"),
        "Cluster IP": spec.get("clusterIP"),
        "Ports": [
            f"{p.get('port')}/{p.get('protocol')} -> {p.get('targetPort')}"
            for p in spec.get("ports", [])
        ],
        "Selector": spec.get("selector"),
    }

def present_secret(data: dict) -> dict:
    """Presents a summary of a Secret."""
    return {
        "Type": data.get("type"),
        "Data Keys": list(data.get("data", {}).keys()),
    }

def present_ingress(data: dict) -> dict:
    """Presents a summary of an Ingress."""
    spec = data.get("spec", {})
    rules = spec.get("rules", [])

    rule_info = [
        {
            "Host": r.get("host"),
            "Paths": [
                f"{p.get('path', '/')} -> {p.get('backend', {}).get('service', {}).get('name')}:{p.get('backend', {}).get('service', {}).get('port', {}).get('number')}"
                for p in r.get("http", {}).get("paths", [])
            ],
        }
        for r in rules
    ]

    return {
        "Ingress Class": spec.get("ingressClassName"),
        "Rules": rule_info,
    }

def present_configmap(data: dict) -> dict:
    """Presents a summary of a ConfigMap."""
    # This is already handled by the special view in the template,
    # but we can return it here for consistency.
    return data.get("data", {})